### DISCUSSION TITLE
Questions about "Help"

### DISCUSSION THREAD

**Discussion by irinaleo** [remkop/picocli#2028] [Q&A] [answered]
I am following along with [this ](https://github.com/remkop/picocli/wiki/JLine-3-Examples#example-for-jline-3141-and-picocli-430---432) example creating an interactive shell. When this is run and you type in "help," it shows three categories of commands: System, Builtins, and PicocliCommands. My only difference from this example is that instead of containing commands in a nested class, all commands are part of the "Example" class as command functions. A few questions:

1. How might I go about changing the name "PicocliCommands" in the help response?
2. Is it possible to only show this category of commands ("PicocliCommands") when help is issued, and perhaps add a -a or --all flag to show the rest?
3. How might I show commands including their options and params using a flag? I would also like to add this to the "help" as an option with -a and --all names

**Comment by irinaleo** [comment]
Regarding my first question, when I create my PicocliCommands instance (picocliCommands), I see that I can print picocliCommands.name(), which is PicocliCommands. I assume this is where the heading comes from in the built-in help subcommand. Can you make a setter for this?

**Reply by remkop** [reply]
> I see that I can print picocliCommands.name() ...

It took me a while to find it, but it looks like the `name()` method is defined in JLine's `CommandRegistry` interface:

```java
// org.jline.console.CommandRegistry
    default String name() {
        return this.getClass().getSimpleName();
    }
```

Yes, we can make a setter, I can see how this would be useful. (I created https://github.com/remkop/picocli/issues/2031 to track this.)
Meanwhile, to change it, please subclass the `PicocliCommands` class in your project and override the `name()` method.

**Comment by remkop** [comment]
Hi @irinaleo thank you for raising this. 

The JLine folks were very helpful in providing the integration layer between picocli and JLine, but I confess JLine is not my specialty. I’ll take a look when I get to my PC.

**Comment by remkop** [comment, marked as answer]
> 1. How might I go about changing the name "PicocliCommands" in the help response?

Until we add a setter for the name (https://github.com/remkop/picocli/issues/2031), please subclass the [PicocliCommands](url) class in your project and override the `name()` method.

> 2. Is it possible to only show this category of commands ("PicocliCommands") when help is issued, and perhaps add a -a or --all flag to show the rest?

I am seeing this line in the Example application:

```java
systemRegistry.setCommandRegistries(builtins, picocliCommands);
```

Omitting the `builtins`, and just registering the picocli commands should do the trick. Can you try this?
```java
systemRegistry.setCommandRegistries(picocliCommands); // without builtins
```

> 3. How might I show commands including their options and params using a flag? I would also like to add this to the "help" as an option with -a and --all names

Ouch, unfortunately I am out of my depth here... This help usage output is largely generated by JLine.
You may need to interact with the JLine maintainers to answer that question.
One alternative could be to avoid JLine's help altogether and only use picocli's help, but that would mean you cannot use the `TailTipWidgets`, and those are kind of nice...  If that is something you want to consider, then please take a look at the [Example in the picocli-shell-jline2 module](https://github.com/remkop/picocli/tree/main/picocli-shell-jline2). Hopefully that can get you started.

**Reply by irinaleo** [reply]
1. Thanks! I am seeing the updated name change so that's certainly a lot more helpful for users of the application who don't (care to) know what picocli is!
2. Omitting it, when I run "help" I still see the heading "Builtins:" and a blank new line below it. So it just removes the commands, not the heading. I think I will keep them though.
3. I will check in with my team and see if the usage is fine as-is. I implemented a separate command "showall" that just prints out `new CommandLine.Help(this).commandList() `. it's just not intuitive for the user to look for such a command.

**Reply by irinaleo** [reply]
@remkop On the topic of `TailTipWidgets`, would you happen to know if I can keep from a suggestion going away? If I type in a valid command, the TailTipWidget shows me the usage until I go to type in a param. It only keeps it displayed if I type in an option like `-h`. I want to keep the usage displayed until I hit 'enter'.

**Reply by remkop** [reply]
@irinaleo Sorry I am not familiar enough with TailTipWidgets to answer that... This is a JLine feature. Can you ask the JLine developers?
